Adds attribute mlstrustedsocket, along with the interface.

qqo opened the pull request:
Hi Tresys!

When building MLS policy, /dev/log socket gets created with just mls_systemhigh sensitivity (s3 in my case, build.conf in [0]), and that contradict mls constraints [1].
I don't quite like setting all of syslogd_t to mlstrustedobject, so the proposed patch does what was discussed in [this thread from 2010](http://oss.tresys.com/pipermail/refpolicy/2010-November/003444.html).
Is this the right direction, or i'm missing something obvious here and should fix that elsewhere?

I didn't touch the "mlsconstrain unix_stream_socket connectto" one, since i haven't seen it in logs yet.


[0] build.conf for reference
```shell
TYPE = mls
NAME = custom
DISTRO = debian
UNK_PERMS = reject
DIRECT_INITRC = n
MONOLITHIC = n
UBAC = y
CUSTOM_BUILDOPT =
MLS_SENS = 4
MLS_CATS = 32
MCS_CATS = 32
QUIET = n
```
[1] sample avc, audit2allow explanation and sesearch.
```shell
root/system_r@testvm:/tmp # grep sendto /var/log/audit/audit.log | tail -1
type=AVC msg=audit(1459979237.878:95): avc:  denied  { sendto } for  pid=1914 comm="login" path="/dev/log" scontext=system_u:system_r:local_login_t:s0-s3:c0.c31 tcontext=system_u:system_r:syslogd_t:s3:c0.c31 tclass=unix_dgram_socket permissive=1
root/system_r@testvm:/tmp # grep sendto /var/log/audit/audit.log | tail -1 | audit2allow -ve 


#============= local_login_t ==============
# audit(1459979237.878:95):
#  scontext="system_u:system_r:local_login_t:s0-s3:c0.c31" tcontext="system_u:system_r:syslogd_t:s3:c0.c31"
#  class="unix_dgram_socket" perms="sendto"
#  comm="login" exe="" path=""
#  message="type=AVC msg=audit(1459979237.878:95): avc:  denied  { sendto } for
#   pid=1914 comm="login" path="/dev/log"
#   scontext=system_u:system_r:local_login_t:s0-s3:c0.c31
#   tcontext=system_u:system_r:syslogd_t:s3:c0.c31 tclass=unix_dgram_socket
#   permissive=1 "
#!!!! This avc is a constraint violation.  You would need to modify the attributes of either the source or target types to allow this access.
#Constraint rule: 
        mlsconstrain unix_dgram_socket { sendto } ((l1 eq l2 -Fail-)  or (t1 == mlsnetwriteranged -Fail-)  and (l1 dom l2 -Fail-)  and (l1 domby h2)  or (t1 == mlsnetwritetoclr -Fail-)  and (h1 dom l2)  and (l1 domby l2)  or (t1 == mlsnetwrite -Fail-)  or (t2 == mlstrustedobject -Fail-) ); Constraint DENIED

#       Possible cause is the source level (s0-s3:c0.c31) and target level (s3:c0.c31) are different.
allow local_login_t syslogd_t:unix_dgram_socket sendto;
root/system_r@testvm:/tmp # sesearch --allow -s local_login_t -t syslogd_t -p sendto
Found 1 semantic av rules:
   allow local_login_t syslogd_t : unix_dgram_socket sendto ; 
```

pebenito commented on policy/mls:
```diff
@@ -216,7 +216,7 @@ mlsconstrain unix_dgram_socket sendto
 	 (( t1 == mlsnetwritetoclr ) and ( h1 dom l2 ) and ( l1 domby l2 )) or
 	 ( t1 == mlsnetwrite ) or
-	 ( t2 == mlstrustedobject ));
+	 ( t2 == mlstrustedsocket ));
```
I'm open to this patch set, but for this constraint change, the t2==mlstrustedsocket should be an additional "or", so existing usages of mlstrustedobject are not broken.

qqo commented:
My bad, replacing the original condition is indeed may break things, I've reverted the original patch and added new check via "or" as you suggested. Should i do this also for network sockets (the constrain just above)? Although I've yet to see the AVC on that one in logs, it looks a bit inconsistent.

Or it would be a better solution to rename this one to mlstrustedunixsocket, for example?

pebenito commented:
I think we should keep it as mlstrustedsocket, as the MLS exemptions are fairly broad (e.g. "file" exemptions also apply to dir, lnk_file, etc.). It is also ready, should a socket with a similar kind of check come along in the future.

It should apply to both unix socket sendto/connectto constraints.

qqo commented:
Noticed that the interface description was misleading, v4 corrects that.
Unfortunately, few white-space-only changes also slipped in, so if thats critical and needs to be redone, please comment, I'll follow up with v5 without them.

pebenito commented:
This looks good, though I'd prefer you squash the patches and remove the whitespace changes
